test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs

```diff
@@ -0,0 +1,107 @@
+            DataFrame df = MakeTestDataFrame(lenght);
+            
+            var grouping = df.GroupBy<string>("Parity").Groupings;
+
+            //Check groups count
+            Assert.Equal(2, grouping.Count());
+
+            //Check number of elements in each group
+            var oddGroup = grouping.Where(gr => gr.Key == "odd").FirstOrDefault();
+            Assert.NotNull(oddGroup);
+            Assert.Equal(lenght/2, oddGroup.Count());
+
+            var evenGroup = grouping.Where(gr => gr.Key == "even").FirstOrDefault();
+            Assert.NotNull(evenGroup);
+            Assert.Equal(lenght / 2 + lenght % 2, evenGroup.Count());
+
+            //Check corner cases
+            lenght = 0;
+            df = MakeTestDataFrame(lenght);
+            grouping = df.GroupBy<string>("Parity").Groupings;
+            Assert.Empty(grouping);
+
+            lenght = 1;
+            df = MakeTestDataFrame(lenght);
+            grouping = df.GroupBy<string>("Parity").Groupings;
+            Assert.Single(grouping);
+            Assert.Equal("even", grouping.First().Key);
+        }
+        
+        [Fact]
+        public void TestGroupingWithTKeyPrimitiveType()
+        {
+            const int lenght = 55;
+
+            //Create test dataframe (numbers starting from 0 up to lenght)
+            DataFrame df = MakeTestDataFrame(lenght);
+
+            //Group elements by int column, that contain the amount of full tens in each int
+            var groupings = df.GroupBy<int>("Tens").Groupings.ToDictionary(g => g.Key, g => g.ToList());
+
+            //Get the amount of all number based columns
+            int numberColumnsCount = df.Columns.Count - 2; //except "Parity" and "Tens" columns
+            
+            //Check each group
+            for (int i = 0; i < lenght / 10; i++)
+            {
+                Assert.Equal(10, groupings[i].Count());
+
+                var rows = groupings[i];
+                for (int colIndex = 0; colIndex < numberColumnsCount; colIndex++)
+                {
+                    var values = rows.Select(row => Convert.ToInt32(row[colIndex]));
+
+                    for (int j = 0; j < 10; j++)
+                    {
+                        Assert.Contains(i * 10 + j, values);
+                    }
+                }
+            }
+
+            //Last group should contain smaller amount of items
+            Assert.Equal(lenght % 10, groupings[lenght / 10].Count());
+        }
+
+        [Fact]
+        public void TestGroupingWithTKeyOfWrongType()
+        {
+            const int lenght = 5;
+
+            var message = string.Empty;
+
+            //Create test dataframe (numbers starting from 0 up to lenght)
+            DataFrame df = MakeTestDataFrame(lenght);
+
+            //Use wrong type for grouping
+            Assert.Throws<InvalidCastException>(() => df.GroupBy<double>("Tens"));
+        }
+
+
+        private DataFrame MakeTestDataFrame(int length)
```

Comment:
While you are here, would you mind renaming this to `MakeTestDataFrameWithParityAndTensColumns`? Just so it's clear what columns are being created.

Reply:
fixed